### PR TITLE
Add test_alpine to circleci config

### LINKED ISSUE #6943: Run CI on Alpine/musl
Since musl is the only way to reliably compile with statically linked libraries, I'd suggest to improve its support level. Mainly, automatic tests would be great.
This shouldn't be very difficult to integrate into CI. Both travis and circle support docker images and the specs can just be executed in an alpine image for `i686-linux-musl` and `x86_64-linux-musl`.

/cc #6903 #6917

### PR DESCRIPTION
**Pull request by straight-shoota**
Fixes #6943

This is just experimental for now.

### CODE REVIEW

**Comment by j8r** [comment]
Great @straight-shoota ! You use my image 😮 
Considering Alpine is often used for static compilation, why not enabling it too here?
Another point, I hope my image will be put in the crystal-lang organization and will become official :)

**Comment by straight-shoota** [comment]
Yes, we'll likely need to use a custom image from crystal org which gets updated on every release build and doesn't have to depend on apk being up to date.

But your image is fine for a preliminary exploration :+1:

**Comment by ysbaddaden** [comment]
You can extract the static crystal binary from the official tarball. Here is a Dockerfile that will configure an Alpine Linux 3.9 image with everything needed to execute a complete workflow on CI:

```Dockerfile
FROM alpine:3.9

ARG VERSION=0.27.2
ARG REVISION=1
ARG ARCH=x86_64
ARG TARBALL=https://github.com/crystal-lang/crystal/releases/download/${VERSION}/crystal-${VERSION}-${REVISION}-linux-${ARCH}.tar.gz

RUN apk add --virtual .build-deps \
        build-base \
        curl \
        git

RUN apk add --virtual .crystal-deps \
        gc-dev \
        libevent-dev \
        libunwind-dev \
        llvm5-dev \
        pcre-dev

RUN apk add --virtual .crystal-specs-deps \
        gmp-dev \
        libxml2-dev \
        openssl-dev \
        readline-dev \
        yaml-dev \
        zlib-dev

RUN set -e; \
    curl -sL ${TARBALL} | tar xz -C /tmp; \
    cp /tmp/crystal-${VERSION}-${REVISION}/lib/crystal/bin/* /usr/local/bin/; \
    rm -rf /tmp/crystal-${VERSION}-${REVISION}
```

**Comment by straight-shoota** [comment]
Unfortunately, the generic linux binary doesn't work on alpine/musl because the baked in gnu target triple (see #7196).

But if it did, that would exactly be the way I've described in https://github.com/crystal-lang/crystal/issues/6943#issuecomment-463179102

**Comment by ysbaddaden** [comment]
@straight-shoota just `export FLAGS=--target $(llvm-config --host-target)`

We can't pass `--target` to `crystal run` *but* the Makefile only calls `crystal build` where we can pass `--target` and it will just work (verified).

**Comment by straight-shoota** [comment]
Yes, macro runs is the main issue that can't be worked around (unless with `sed -i -e 's/x86_64-unknown-linux-gnu/x86_64-alpine-linux-musl/' /usr/local/bin/crystal`). The default target should not be baked in but depend on `ldd --version` (or `llvm-config --host-target`), at least to distinguish between linux-musl and linux-gnu.

**Comment by ysbaddaden** [comment]
It would be great to have a minimal bootstrap compiler with just `build` command. No playground tools, or docs, not even run or eval. ping @asterite

**Comment by asterite** [comment]
@ysbaddaden Yes, I always wanted to move those tools to separate binaries, even if it means having 4 or 5 "big" binaries (right now crystal is about 42MB, so maybe all of them would occupy at most 200MB, maybe less because each binary will be smaller). I'd really like that to happen. Also removing some parts of the std to separate shards (that could live in the `crystal-lang` organization, like XML, YAML, OAuth, etc.). But last thing I remember is that most didn't want to go into that direction (one thing was avoiding a more complex build/release that would involve multiple binaries, I don't think that's a big issue given how automated things are now).

**Comment by j8r** [comment]
That's already nice to have moved out package management. I've worked to move out Crystal playground, apparently only few wanted to. I would work again on it if there is a consensus on it.

**Comment by straight-shoota** [comment]
What does this all have to do with ci builds on alpine?

**Comment by asterite** [comment]
I think nothing 😊

**Comment by j8r** [comment]
I think because i complicates portability and porting on other environments.

**Comment by ysbaddaden** [comment]
@straight-shoota a bootstrap compiler would allow to use the released static crystal binary on alpine, and overall ease cross platform development and experimentations :)

**Comment by straight-shoota** [comment]
I don't see why that would make a difference. We just need to make the target host configurable for *all* compilation steps (thus including macro evaluation) and use a default depending on the environment. That's the same for both a small as well as a fat compiler.

**Comment by asterite** [comment]
Macros run in the host, why would that need to be configurable?

**Comment by straight-shoota** [comment]
Configurable as in "configures itself", instead of being baked in. So the same compiler binary can work with both musl and gnu by default (and without requiring any manual configuration).

**Comment by straight-shoota** [comment]
Now the specs are running into `Failed to raise an exception: END_OF_STACK` https://circleci.com/gh/crystal-lang/crystal/19693?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Maybe this is related to #6934? Although it's not a static build.

**Comment by j8r** [comment]
@straight-shoota do you still work on this? The CI can be re-triggered, mps has pushed Crystal 0.29 to Alpine, my image use this version.
This CI will be very useful to prevent issues like [this one](https://github.com/crystal-lang/crystal/pull/7871).

**Comment by straight-shoota** [comment]
@j8r I've triggered a rebuild but it still fails for some reason I'm not going to investigate further. This is also an old branch which would need a rebase.

But anyway, this should build on crystal-lang/distribution-scripts#29 for a custom alpine-based Dockerfile so we don't need to rely on alpine package updates (which might take a few days after a Crystal release).

**Comment by j8r** [comment]
Indeed, the image will have to change. I've removed g++ on the image, because only the Crystal compiler needs it. That's offtopic: why g++ over Clang?

**Comment by j8r** [comment]
Any plan to have an Alpine CI any time soon?
I experimented [with Drone CI](https://github.com/crystal-lang/crystal/issues/8080#issuecomment-539579581), but it looks like the tests are too long to finish (cancelled because >1hour).

**Comment by straight-shoota** [comment]
I'd like to move forward, but it is held up by crystal-lang/distribution-scripts#29 Waiting for feedback there.

**Comment by straight-shoota** [comment]
This depends on crystal-lang/distribution-scripts#55 which is already used for building `crystallang/crystal:ci-test-dev` docker image.
There's one failure in stdlib specs: #7482

**Comment by bcardiff** [comment]
It would be great to emit the junit output as other jobs. The `*ci_steps` does not have those instructions, it can be added there. 

NB: the *ci_steps is currently not used due to the workaround of splitting the specs. But let's improveit and keep it. The DRY will come back at some point, will come back at some point.

**Comment by j8r** [comment]
Just a thought: if Alpine aarch64 CI is planned, wouldn't be better to have all Alpine CIs in GH Actions?
Since it is free, perhaps it would save some hours credits from Circle CI (I don't know the billing plan).

**Comment by straight-shoota** [comment]
There are currently no plans for Alpine aarch64. At least I'm not aware of any. Duplicating the workflow on circleci is far easier integration than setting up a new GitHub action. And AFAIK circleci credits are not a concern atm.

So we should stick with this PR which is ready to merge. The important goal is to gain spec coverage linking against musl-libc. We can always reconsider whenever we get around implementing CI for alpine on aarch64. But that's currently not a high-priority concern. Maybe at that point we might even have advanced further towards a single, integrated CI platform.

File: bin/ci [hunk 1]
```diff
@@ -148,6 +148,9 @@ with_build_env() {
     x86_64)
       export DOCKER_TEST_IMAGE="$DOCKER_TEST_PREFIX-build"
       ;;
+    x86_64-musl)
```
Comment (bcardiff) [inline bin/ci]: Shouldn't it be the other way around to mimic the target triple?

```suggestion
    musl-x86_64)
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -62,6 +71,16 @@ jobs:
...
           path: ~/test-results/spec
 
+  test_alpine:
+    machine: true
+    environment:
+      <<: *env
+      TRAVIS_OS_NAME: linux
+      ARCH: x86_64-musl
+      ARCH_CMD: linux64
+      DOCKER_TEST_PREFIX: crystallang/crystal:ci-test-dev # TODO: Remove after 0.33.0
```
Comment (bcardiff) [inline .circleci/config.yml]: I'm pushing a `0.32.1-alpine[-build]` with the changes of https://github.com/crystal-lang/distribution-scripts/pull/55 included. So this line can be removed and we can verify how it works.

File: bin/ci (hunk 1 above)
Comment (straight-shoota) [reply bin/ci]: How does that mimic? The target triples are `x86_64-linux-gnu` and `x86_64-linux-musl`.

Comment (bcardiff) [reply bin/ci]: I got confused with the name of the .tar.gz which are *-linux-x86_64.

**Comment by straight-shoota** [comment]
I've removed the explicit config, now the job uses the regular alpine image. This is ready to merge!

**Pull request merged**